Tiltfile

```diff
@@ -186,6 +186,8 @@ def enable_provider(name):
     for substitution in substitutions:
         value = substitutions[substitution]
         yaml = yaml.replace("${" + substitution + "}", value)
+    if len(substitutions)==0:
```

Comment:
I suggest to check if there are leftover variables in the yaml instead of checking if there are at least one variable defined in tilt_settings.json.
eg. I have AWS_B64ENCODED_CREDENTIALS defined, but I switch to use VSPHERE. Current test will pass, but the management cluster won't work because VSPHERE_USERNAME and VSPHERE_PASSWORD variables are missing.

FYI, the regex used for searching variables in clusterctl is ${\s*([A-Z0-9_]+)\s*}

Reply:
didn't think to deeply about this - was just a drive by. 
do you feel strongly about this? If so i can dive back in and check on it in a few days.  

If any other improvments i can bundle in a broader patch ?

Reply:
I don't think we should expand the scope.
I like your idea of warning for missing variables replacement (I stumbled in the same problem several times), I'm just advocating for a better definition of "missing variable" 😉